File: src/test/java/io/gaegul/util/ConsoleIOTest.java
package io.gaegul.util;

import java.io.ByteArrayOutputStream;
import java.io.PrintStream;

import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;

/**
 * 콘솔 출력 문자열 테스트 헬퍼 클래스
 */
public abstract class ConsoleIOTest {

	private ByteArrayOutputStream outputStreamCaptor;
	private PrintStream standardOut; // 표준 스트림

	@BeforeEach
	void setUp() {
		standardOut = System.out; // 표준 스트림 초기화
		outputStreamCaptor = new ByteArrayOutputStream();
		System.setOut(new PrintStream(outputStreamCaptor));
	}

	@AfterEach
	protected void printResult() {
		System.setOut(standardOut); // 표준 스트림 할당
		System.out.println(output()); // 원하는 내용이 잘 나왔는지 문자열 디코딩 바이트를 가져와 출력
	}

	protected String output() {
		return outputStreamCaptor.toString().trim();
	}
}

File: src/test/kotlin/io/gaegul/utils/ConsoleIOTest.kt
package io.gaegul.utils

import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import java.io.ByteArrayOutputStream
import java.io.PrintStream

/**
 * 콘솔 출력 문자열 테스트 헬퍼 클래스
 */
abstract class ConsoleIOTest {
    lateinit var outputStreamCaptor: ByteArrayOutputStream
    lateinit var standardOut: PrintStream // 표준 스트림 초기화

    @BeforeEach
    fun setUp() {
        /* 표준 스트림 초기화 */
        standardOut = System.out
        outputStreamCaptor = ByteArrayOutputStream()
        System.setOut(PrintStream(outputStreamCaptor))
    }

    @AfterEach
    fun tearDown() {
        /* 표준 스트림 할당 */
        System.setOut(standardOut)
        /* 원하는 내용이 잘 나왔는지 문자열 디코딩 바이트를 가져와 출력 */
        println(output())
    }

    fun output(): String {
        return outputStreamCaptor.toString().trim()
    }
}